src/transformers/models/propainter/modeling_propainter.py

```diff
@@ -0,0 +1,4620 @@
+        )
+
+    def forward(self, network, input, correlation, optical_flow):
+        motion_features = self.encoder(optical_flow, correlation)
+        input = torch.cat([input, motion_features], dim=1)
+
+        network = self.gru(network, input)
+        delta_flow = self.flow_head(network)
+        # scale mask to balance gradients
+        mask = 0.25 * self.mask(network)
+        return network, mask, delta_flow
+
+
+def coords_grid(batch_size, height, width):
+    coords = torch.meshgrid(torch.arange(height), torch.arange(width))
+    coords = torch.stack(coords[::-1], dim=0).float()
+    return coords[None].repeat(batch_size, 1, 1, 1)
+
+
+def sample_point(img, coords):
+    """Wrapper for grid_sample, uses pixel coordinates"""
+    height, width = img.shape[-2:]
+    xgrid, ygrid = coords.split([1, 1], dim=-1)
+    xgrid = 2 * xgrid / (width - 1) - 1
+    ygrid = 2 * ygrid / (height - 1) - 1
+
+    grid = torch.cat([xgrid, ygrid], dim=-1)
+    img = F.grid_sample(img, grid, align_corners=True)
+
+    return img
+
+
+class ProPainterCorrBlock:
+    def __init__(
+        self,
+        config: ProPainterConfig,
+        feature_map_1: torch.tensor,
+        feature_map_2: torch.tensor,
+        num_levels: int = 4,
+        radius: int = 4,
+    ):
+        self.config = config
+        self.num_levels = num_levels
+        self.radius = radius
+        self.correlation_pyramid = []
+
+        # all pairs correlation
+        correlation = ProPainterCorrBlock.correlation(feature_map_1, feature_map_2)
+
+        batch_size, height_1, width_1, dimension, height_2, width_2 = correlation.shape
+        correlation = correlation.reshape(batch_size * height_1 * width_1, dimension, height_2, width_2)
+
+        self.correlation_pyramid.append(correlation)
+        for _ in range(self.num_levels - 1):
+            correlation = F.avg_pool2d(correlation, 2, stride=config.multi_level_conv_stride[1])
+            self.correlation_pyramid.append(correlation)
+
+    def __call__(self, coords):
+        radius = self.radius
+        coords = coords.permute(0, 2, 3, 1)
+        batch_size, height_1, width_1, _ = coords.shape
+
+        out_pyramid = []
+        for i in range(self.num_levels):
+            correlation = self.correlation_pyramid[i]
+            delta_x = torch.linspace(-radius, radius, 2 * radius + 1)
+            delta_y = torch.linspace(-radius, radius, 2 * radius + 1)
+            delta = torch.stack(torch.meshgrid(delta_y, delta_x), axis=-1).to(coords.device)
+            centroid_lvl = coords.reshape(batch_size * height_1 * width_1, 1, 1, 2) / 2**i
+            delta_lvl = delta.view(1, 2 * radius + 1, 2 * radius + 1, 2)
+            coords_lvl = centroid_lvl + delta_lvl
+            correlation = sample_point(correlation, coords_lvl)
+            correlation = correlation.view(batch_size, height_1, width_1, -1)
+            out_pyramid.append(correlation)
```

Comment:
minor nit, and a general suggestion on naming explicitly variables

```suggestion
        for level in range(self.num_levels):
            correlation = self.correlation_pyramid[level]
            delta_x = torch.linspace(-radius, radius, 2 * radius + 1)
            delta_y = torch.linspace(-radius, radius, 2 * radius + 1)
            delta = torch.stack(torch.meshgrid(delta_y, delta_x), axis=-1).to(coords.device)
            centroid_level = coords.reshape(batch_size * height_1 * width_1, 1, 1, 2) / 2**level
            delta_level = delta.view(1, 2 * radius + 1, 2 * radius + 1, 2)
            coords_level = centroid_level + delta_level
            correlation = sample_point(correlation, coords_level)
            correlation = correlation.view(batch_size, height_1, width_1, -1)
            out_pyramid.append(correlation)
```